File: lib/streaming/session-manager.ts
/**
 * Session management for SSE connections
 */

import { v4 as uuidv4 } from 'uuid';

export interface Session {
  id: string;
  createdAt: Date;
  lastActivity: Date;
  isActive: boolean;
  metadata?: Record<string, any>;
}

export class SessionManager {
  private sessions: Map<string, Session> = new Map();
  private activeSessionId: string | null = null;
  
  /**
   * Create a new session
   */
  createSession(metadata?: Record<string, any>): Session {
    const session: Session = {
      id: uuidv4(),
      createdAt: new Date(),
      lastActivity: new Date(),
      isActive: true,
      metadata
    };
    
    this.sessions.set(session.id, session);
    this.activeSessionId = session.id;
    
    return session;
  }
  
  /**
   * Get a session by ID
   */
  getSession(sessionId: string): Session | undefined {
    return this.sessions.get(sessionId);
  }
  
  /**
   * Get the active session
   */
  getActiveSession(): Session | null {
    if (!this.activeSessionId) return null;
    return this.sessions.get(this.activeSessionId) || null;
  }
  
  /**
   * Update session activity
   */
  updateActivity(sessionId: string): void {
    const session = this.sessions.get(sessionId);
    if (session) {
      session.lastActivity = new Date();
    }
  }
  
  /**
   * Set active session
   */
  setActiveSession(sessionId: string): boolean {
    if (this.sessions.has(sessionId)) {
      this.activeSessionId = sessionId;
      return true;
    }
    return false;
  }
  
  /**
   * Deactivate a session
   */
  deactivateSession(sessionId: string): void {
    const session = this.sessions.get(sessionId);
    if (session) {
      session.isActive = false;
      if (this.activeSessionId === sessionId) {
        this.activeSessionId = null;
      }
    }
  }
  
  /**
   * Remove a session
   */
  removeSession(sessionId: string): boolean {
    if (this.activeSessionId === sessionId) {
      this.activeSessionId = null;
    }
    return this.sessions.delete(sessionId);
  }
  
  /**
   * Clean up inactive sessions older than maxAge
   */
  cleanupSessions(maxAgeMs: number = 3600000): number {
    const now = Date.now();
    let removed = 0;
    
    this.sessions.forEach((session, id) => {
      const age = now - session.lastActivity.getTime();
      if (!session.isActive && age > maxAgeMs) {
        this.removeSession(id);
        removed++;
      }
    });
    
    return removed;
  }
  
  /**
   * Get all active sessions
   */
  getActiveSessions(): Session[] {
    return Array.from(this.sessions.values()).filter(s => s.isActive);
  }
  
  /**
   * Clear all sessions
   */
  clearAll(): void {
    this.sessions.clear();
    this.activeSessionId = null;
  }
}

// Singleton instance
export const sessionManager = new SessionManager();

File: backend/src/utils/session_manager.py
"""Simple session management for POC."""

import asyncio
import logging
import time
from dataclasses import dataclass, field
from typing import Any

logger = logging.getLogger(__name__)


@dataclass
class SessionInfo:
    """Session metadata."""

    session_id: str
    user_id: str
    created_at: float = field(default_factory=time.time)
    last_activity: float = field(default_factory=time.time)
    request_queue: Any = None  # LiveRequestQueue
    metadata: dict[str, Any] = field(default_factory=dict)

    def update_activity(self):
        """Update last activity timestamp."""
        self.last_activity = time.time()

    @property
    def age_seconds(self) -> float:
        """Get session age in seconds."""
        return time.time() - self.created_at

    @property
    def inactive_seconds(self) -> float:
        """Get time since last activity in seconds."""
        return time.time() - self.last_activity


class SessionManager:
    """Simple in-memory session manager for POC."""

    def __init__(self, max_age_seconds: int = 3600):  # 1 hour default
        self.sessions: dict[str, SessionInfo] = {}
        self.max_age_seconds = max_age_seconds
        self._cleanup_task: asyncio.Task | None = None
        self._running = False

    async def start(self):
        """Start the session manager with periodic cleanup."""
        if not self._running:
            self._running = True
            self._cleanup_task = asyncio.create_task(self._periodic_cleanup())
            logger.info(f"Session manager started with {self.max_age_seconds}s max age")

    async def stop(self):
        """Stop the session manager."""
        self._running = False
        if self._cleanup_task:
            self._cleanup_task.cancel()
            try:
                await self._cleanup_task
            except asyncio.CancelledError:
                pass
        logger.info("Session manager stopped")

    def create_session(
        self, session_id: str, user_id: str, request_queue: Any = None
    ) -> SessionInfo:
        """Create a new session."""
        session = SessionInfo(
            session_id=session_id, user_id=user_id, request_queue=request_queue
        )
        self.sessions[session_id] = session
        logger.info(f"Session created: {session_id} for user {user_id}")
        return session

    def get_session(self, session_id: str) -> SessionInfo | None:
        """Get session by ID and update activity."""
        session = self.sessions.get(session_id)
        if session:
            session.update_activity()
        return session

    def remove_session(self, session_id: str) -> SessionInfo | None:
        """Remove and return a session."""
        session = self.sessions.pop(session_id, None)
        if session:
            # Clean up request queue if exists
            if session.request_queue:
                session.request_queue.close()
            logger.info(f"Session removed: {session_id}")
        return session

    def get_session_readonly(self, session_id: str) -> SessionInfo | None:
        """Get session by ID without updating activity (for debugging)."""
        return self.sessions.get(session_id)

    def get_active_session_count(self) -> int:
        """Get count of active sessions."""
        return len(self.sessions)

    async def _periodic_cleanup(self):
        """Periodically clean up expired sessions."""
        cleanup_interval = min(
            300, max(self.max_age_seconds / 4, 1)
        )  # Check every 5 min or 1/4 of max age, with a minimum of 1 second

        while self._running:
            try:
                await asyncio.sleep(cleanup_interval)

                # Find expired sessions
                expired = []
                for session_id, session in self.sessions.items():
                    if session.age_seconds > self.max_age_seconds:
                        expired.append(session_id)

                # Clean up expired sessions
                for session_id in expired:
                    self.remove_session(session_id)

                if expired:
                    logger.info(f"Cleaned up {len(expired)} expired sessions")

                # Log status
                active_count = self.get_active_session_count()
                if active_count > 0:
                    logger.debug(f"Active sessions: {active_count}")

            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Error in session cleanup: {e}")


# Global session manager instance
session_manager = SessionManager()
